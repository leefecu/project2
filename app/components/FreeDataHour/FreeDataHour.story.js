import React from 'react';
import { View } from 'react-native';

import { storiesOf } from '@storybook/react-native';
import { action } from '@storybook/addon-actions';

import FreeDataHour from './FreeDataHour';
import Styles from '../../Stories.styles';

storiesOf('Component | FreeDataHour', module).add('Default', () => {
  return (
    <View style={Styles.containerWithBG}>
      <View style={Styles.row}>
        <FreeDataHour onStart={action('onStart')} title="Free Data Hour" />
      </View>
    </View>
  );
});
