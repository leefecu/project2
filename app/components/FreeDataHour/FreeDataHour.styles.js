import { StyleSheet } from 'react-native';

import Colors from '../../themes/Colors';
import Metrics from '../../themes/Metrics';

export const FREE_DATA_HOUR_WIDTH = Metrics.width - Metrics.doublePadding * 2;

export default StyleSheet.create({
  container: {
    alignItems: 'stretch',
    flex: 1,
    flexDirection: 'column',
    position: 'relative',
    width: FREE_DATA_HOUR_WIDTH,
  },
  iconRow: {
    backgroundColor: Colors.transparent,
    height: 36,
    position: 'relative',
    zIndex: 1,
  },
  icon: {
    position: 'absolute',
    top: 0,
    left: FREE_DATA_HOUR_WIDTH / 2 - 27,
  },
  labelRow: {
    backgroundColor: Colors.white,
    alignItems: 'center',
    justifyContent: 'center',
    paddingTop: Metrics.largePadding,
    paddingBottom: Metrics.basePadding,
  },
  buttonRow: {
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.sunriseYellow,
    paddingVertical: Metrics.basePadding,
    paddingHorizontal: Metrics.doublePadding,
  },
  label: {
    color: Colors.primaryDark,
    fontSize: Metrics.Fonts.normal,
    fontFamily: 'CircularPro-Bold',
  },
});
