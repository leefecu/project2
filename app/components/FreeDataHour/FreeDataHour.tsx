import React from 'react';
import { Text, View } from 'react-native';

import Button from '../Button/Button';

import Styles, { FREE_DATA_HOUR_WIDTH } from './FreeDataHour.styles';
import ClockSVG from '../../images/clockOn.svg';
import Metrics from '../../themes/Metrics';

interface Props {
  onStart: () => void;
  title: string;
}

const BUTTON_WIDTH = FREE_DATA_HOUR_WIDTH - Metrics.doublePadding * 2;

const FreeDataHour = ({ onStart, title }: Props) => {
  return (
    <View style={Styles.container}>
      <View style={Styles.iconRow}>
        <ClockSVG style={Styles.icon} />
      </View>
      <View style={Styles.labelRow}>
        <Text style={Styles.label}>{title}</Text>
      </View>
      <View style={Styles.buttonRow}>
        <Button
          onPress={onStart}
          id={`Start Free Data Hour ${title}`}
          label='Start the clock'
          width={BUTTON_WIDTH}
        />
      </View>
    </View>
  );
};

export default FreeDataHour;
