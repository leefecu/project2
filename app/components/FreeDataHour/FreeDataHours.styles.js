import { StyleSheet } from 'react-native';

import Metrics from '../../themes/Metrics';

export default StyleSheet.create({
  packRow: {
    marginBottom: Metrics.baseMargin,
  },
});
