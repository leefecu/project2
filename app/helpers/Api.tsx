import get from 'lodash/get';
import reject from 'lodash/reject';
import Config from 'react-native-config';
import compareVersions from 'compare-versions';
import { Platform } from 'react-native';

import * as Storage from './Storage';
import { createExpiryNotification } from './Notification';
import { convertStashToActivePAck } from './Transformation';
import API from '../services/Api';

import { SYSTEM_RESPONSE } from '../constants/Shared';

const api = API.create();

export const purchasePack = async (packetId: string, promoId?: string) => {
  const response = await api.purchasePack(packetId, promoId);
  const currentActivePack = await Storage.getActivePack();
  if (response.ok) {
    const balance = get(response, 'data.balance');
    const purchasedPack = get(response, 'data.purchasedPack');
    const activePack = currentActivePack
      ? {
          ...currentActivePack,
          endTime: purchasedPack.endTime,
        }
      : purchasedPack;

    createExpiryNotification(purchasedPack.endTime);
    await Storage.setActivePackAndBalance(activePack, balance);
    return response;
  }
  return response;
};

export const activateStash = async (stash: TStash) => {
  const response = await api.activateStash(stash.packetInstanceId);
  if (response.ok) {
    const currentActivePack = await Storage.getActivePack();
    const currentStash = await Storage.getStash();
    const newActivePack = convertStashToActivePAck(stash, currentActivePack);
    const newStash = reject(
      currentStash,
      (target: TStash) => target.packetInstanceId === stash.packetInstanceId,
    );

    createExpiryNotification(newActivePack.endTime);
    await Storage.saveActivePackAndStash(newActivePack, newStash);
    return response;
  }
  return response;
};

export const checkSystemInfo = async (): Promise<SYSTEM_RESPONSE> => {
  const response = await api.getSystemInfo();
  if (response.ok) {
    const maintenance: boolean = get(response, 'data.maintenance', false);
    if (maintenance) {
      return SYSTEM_RESPONSE.MAINTENANCE;
    }
    const minVersion = get(response, `data.${Platform.OS}`);
    const appVersion = Config.APP_VERSION;

    return compareVersions.compare(appVersion, minVersion, '>=')
      ? SYSTEM_RESPONSE.CONTINUE
      : SYSTEM_RESPONSE.UPGRADE;
  }
  return SYSTEM_RESPONSE.CONTINUE;
};
