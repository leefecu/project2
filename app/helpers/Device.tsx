import { NativeModules, Platform } from 'react-native';
import DeviceInfo from 'react-native-device-info';
import CarrierInfo from 'react-native-carrier-info';
import RNFS from 'react-native-fs';

const NZ_MCC = '530';
const NZ_COUNTRY_CODE = 'nz';
const IS_IOS = Platform.OS === 'ios';

const NetworkInfo = NativeModules.RNNetworkInfo;

export const isSimulator = async (): Promise<boolean> => {
  const isEmulator = await DeviceInfo.isEmulator();
  return isEmulator;
};

export const isSimIn = async () => {
  // react-native-carrier-info does not work in simulator
  if (await isSimulator()) {
    return true;
  }

  try {
    // mobileCountryCode method fails when no SIM card in the device
    await CarrierInfo.mobileCountryCode();
    return true;
  } catch (e) {
    return false;
  }
};

export const checkForRoaming = async (debug: boolean = false): Promise<boolean> => {
  // react-native-carrier-info does not work in simulator
  if (await isSimulator()) {
    return debug;
  }

  try {
    if (Platform.OS === 'ios') {
      const operatorPlist = await NativeModules.NetworkInfo.getOperatorPlist();
      const mcc = operatorPlist.match(/\d+\/carrier.plist/g);
      if (mcc && mcc.length > 0) {
        return mcc[0].substr(0, 3) !== NZ_MCC;
      }
      return false;
    }
    const countryCode = await NetworkInfo.isoCountryCode();
    return countryCode !== NZ_COUNTRY_CODE;
  } catch (e) {
    return false;
  }
};

const majorVersionIOS: number = parseInt(`${Platform.Version}`, 10);

export const isAutoFillSupported: boolean = IS_IOS && majorVersionIOS >= 12;

// Figured out how current(old) DataClock app stores msisdn
// it's stored in xml file
// shared_prefs/nz.co.twodegreesmobile.comptelfwd.smpl.Account.xml
const msisdnFileName = 'nz.co.twodegreesmobile.comptelfwd.smpl.Account.xml';
const msisdnFolderName = 'shared_prefs';
const msisdnRegex = /com.comptel.smpl.interaction.Account.PHONE_NUMBER_KEY">\d+<\/string/g;

export const getExistingMsisdn = async (): Promise<TMigrationData> => {
  if (IS_IOS) {
    try {
      const msisdn = await NativeModules.GetNumber.getMsisdn();
      return { number: msisdn, upgrade: !!msisdn };
    } catch (e) {
      return { number: '', upgrade: false };
    }
  }
  // This method only works with Android
  try {
    // e.g. RNFS.DocumentDirectoryPath = '/data/user/0/nz.co.twodegreesmobile.comptelfwd.smpl/files',
    const paths = RNFS.DocumentDirectoryPath.split('/');
    // need to remove "/files" and replace with "shared_prefs"
    paths[paths.length - 1] = msisdnFolderName;
    const path = paths.join('/');
    const xmlPath = `${path}/${msisdnFileName}`;
    const existingAccount = await RNFS.readFile(xmlPath);
    const match = msisdnRegex.exec(existingAccount);
    if (match && match.length > 0) {
      const msisdn = match[0]
        .replace('com.comptel.smpl.interaction.Account.PHONE_NUMBER_KEY">', '')
        .replace('</string', '');

      return { number: msisdn, upgrade: true };
    }
    return { number: '', upgrade: false };
  } catch (e) {
    return { number: '', upgrade: false };
  }
};
