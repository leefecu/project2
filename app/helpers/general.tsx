import { NavigationParams, NavigationScreenProp, NavigationState } from 'react-navigation';

import { ERROR_CODE } from '../constants/Messages';
import { SYSTEM_RESPONSE } from '../constants/Shared';

import * as API from '../helpers/Api';

export const noop = () => {};

interface TTestProp {
  testID: string;
  accessibilityLabel: string;
}
export const testProps = (id: string): TTestProp => ({
  testID: id,
  accessibilityLabel: id,
});

export const shouldLogout = (status: number | undefined, errorCode: ERROR_CODE): boolean =>
  status === 401 ||
  status === 403 ||
  errorCode === ERROR_CODE.NOT_REGISTERED ||
  errorCode === ERROR_CODE.INVALID_PLAN ||
  errorCode === ERROR_CODE.QD_ISSUE;

export const systemCheck = async (
  navigation: NavigationScreenProp<NavigationState, NavigationParams>,
) => {
  const systemInfo = await API.checkSystemInfo();
  if (systemInfo === SYSTEM_RESPONSE.CONTINUE) {
    return;
  }
  navigation.navigate(systemInfo === SYSTEM_RESPONSE.MAINTENANCE ? 'Maintenance' : 'Upgrade');
};
