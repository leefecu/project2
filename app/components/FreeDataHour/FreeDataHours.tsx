import React, { PureComponent } from 'react';
import { FlatList, View } from 'react-native';
import throttle from 'lodash/throttle';

import FreeDataHour from './FreeDataHour';

import { THROTTLING } from '../../constants/Shared';

import { noop } from '../../helpers/general';

import Styles from './FreeDataHours.styles';

interface Props {
  freeDataHours: Array<TFreeDataHour>;
  onSelectPack: (pack: TFreeDataHour) => void;
}

class FreeDataHours extends PureComponent<Props> {
  static defaultProps: Props = {
    freeDataHours: [],
    onSelectPack: noop,
  };

  handleSelectPackThrottled: Function;

  constructor(props: Props) {
    super(props);

    this.handleSelectPackThrottled = throttle(this.handleSelectPack, THROTTLING, {
      trailing: false,
    });
  }

  keyExtractor = (item: TFreeDataHour) => item.promoId;

  handleSelectPack = (item: TFreeDataHour) => {
    const { onSelectPack } = this.props;
    onSelectPack(item);
  };

  renderFreeDataHour = ({ item }: { item: TFreeDataHour }) => {
    return (
      <View style={Styles.packRow}>
        <FreeDataHour onStart={() => this.handleSelectPackThrottled(item)} title={item.title} />
      </View>
    );
  };

  render() {
    const { freeDataHours } = this.props;
    return (
      <FlatList
        data={freeDataHours}
        keyExtractor={this.keyExtractor}
        renderItem={this.renderFreeDataHour}
      />
    );
  }
}

export default FreeDataHours;
