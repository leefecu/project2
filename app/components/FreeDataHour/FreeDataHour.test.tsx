/**
 * @jest-environment jsdom
 */
import React from 'react';
import { create } from 'react-test-renderer';

import FreeDataHour from './FreeDataHour';

import { noop } from '../../helpers/general';

describe('FreeDataHour component', () => {
  test('render default', () => {
    const button = create(<FreeDataHour onStart={noop} title="Free Data Hour" />);
    expect(button.toJSON()).toMatchSnapshot();
  });
});
