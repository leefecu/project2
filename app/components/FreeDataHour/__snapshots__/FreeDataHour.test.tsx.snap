// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FreeDataHour component render default 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flex": 1,
      "flexDirection": "column",
      "position": "relative",
      "width": 702,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 36,
        "position": "relative",
        "zIndex": 1,
      }
    }
  >
    <SVGMock
      style={
        Object {
          "left": 324,
          "position": "absolute",
          "top": 0,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "justifyContent": "center",
        "paddingBottom": 16,
        "paddingTop": 36,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#004071",
          "fontFamily": "CircularPro-Bold",
          "fontSize": 20,
        }
      }
    >
      Free Data Hour
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffce61",
        "justifyContent": "center",
        "paddingHorizontal": 24,
        "paddingVertical": 16,
      }
    }
  >
    <View
      accessibilityLabel="Start Free Data Hour Free Data Hour button"
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "width": 654,
          },
          Object {
            "alignItems": "stretch",
            "backgroundColor": "transparent",
            "borderColor": "#004071",
            "borderRadius": 32,
            "borderWidth": 1,
            "paddingVertical": 12,
          },
          false,
          false,
          false,
          false,
          false,
        ]
      }
      testID="Start Free Data Hour Free Data Hour button"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#004071",
              "fontFamily": "CircularPro-Bold",
              "fontSize": 20,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            false,
            false,
            false,
            false,
            false,
          ]
        }
      >
        Start the clock
      </Text>
    </View>
  </View>
</View>
`;
